rust_src/src/buffers.rs

```diff
@@ -1057,6 +1060,52 @@ pub fn erase_buffer() {
     }
 }
 
+// We split this away from generate-new-buffer, because rename-buffer
+// and set-visited-file-name ought to be able to use this to really
+// rename the buffer properly.
+
+/// Return a string that is the name of no existing buffer based on NAME.
+/// If there is no live buffer named NAME, then return NAME.
+/// Otherwise modify name by appending `<NUMBER>', incrementing NUMBER
+/// (starting at 2) until an unused name is found, and then return that name.
+/// Optional second argument IGNORE specifies a name that is okay to use (if
+/// it is in the sequence to be tried) even if a buffer with that name exists.
+///
+/// If NAME begins with a space (i.e., a buffer that is not normally
+/// visible to users), then if buffer NAME already exists a random number
+/// is first appended to NAME, to speed up finding a non-existent buffer.
+#[lisp_fn(min = "1")]
+pub fn generate_new_buffer_name(name: LispStringRef, ignore: LispObject) -> LispStringRef {
+    if (ignore != Qnil && string_equal(name.into(), ignore)) || Fget_buffer(name.into()).is_nil() {
+        return name;
+    }
+
+    let basename = if name.byte_at(0) == b' ' {
+        let range = Range::new(0, 1_000_000);
+        let mut rng = rand::thread_rng();
+        let mut s = format!("-{}", range.ind_sample(&mut rng));
+        local_unibyte_string!(suffix, s);
+        let genname = unsafe { concat2(name.into(), suffix) };
+        if get_buffer(LispBufferOrName::Name(genname)).is_none() {
+            return genname.into();
+        }
+        genname
+    } else {
+        name.into()
+    };
+
+    let mut suffix_count = 2;
+    loop {
+        let mut s = format!("<{}>", suffix_count);
+        local_unibyte_string!(suffix, s);
+        let candidate = unsafe { concat2(basename, suffix) };
+        if string_equal(candidate, ignore) || Fget_buffer(candidate).is_nil() {
```

Comment:
I think you can use `get_buffer` here.

Reply:
```suggestion
        if string_equal(candidate, ignore) || get_buffer(LispBufferOrName::Name(candidate)).is_none() {
```

Reply:
Since you know you have a string we can be more direct.

Reply:
Done